### PR TITLE
feat(outcome): Move event signals to outcomes consumer

### PR DESCRIPTION
**Pull request by RaduW**
Moves event signal emission from the store endpoint code to a consumer of related outcomes.

When an event is filtered or rate limited ("dropped"), we emit respective signals. Plugins can hook those signals and run callbacks on these actions. In the future, event ingestion will be moved into Relay, which means signals cannot be emitted directly anymore.

To solve this, signals are now emitted for the `FILTERED` and `RATE_LIMITED` outcomes. During migration, we use a key in the Django cache (i.e. memcache) to detect if signals have been emitted already.

### CODE REVIEW

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [inline src/sentry/outcome_consumer.py]: _nit_: Use `outcome in (Outcome.FILTERED, Outcome.RATE_LIMITED)` here.

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [inline src/sentry/outcome_consumer.py]: _nit_: Since project ids can never be zero, you could fold this together:
`project_id = int(msg.get("project_id", 0))`

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [inline src/sentry/outcome_consumer.py]: Please log an error here. This is a condition that should never happen and might indicate misconfiguration or a serious bug.

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [inline src/sentry/outcome_consumer.py]: _nit_: Please rename the variable to `outcomes_consumer`.

File: src/sentry/utils/kafka.py
Comment (jan-auer) [inline src/sentry/utils/kafka.py]: _nit_: Can we use `self.consumer_configuration.update(**(consumer_configuration or {}))` here?

File: src/sentry/utils/kafka.py
Comment (jan-auer) [inline src/sentry/utils/kafka.py]: Please use `sentry.utils.safe_execute` here. This will run a callback in a try-catch block and log the error appropriately.

File: src/sentry/utils/kafka.py
Comment (jan-auer) [inline src/sentry/utils/kafka.py]: Looking at this again now, should we maybe just log the error and continue with the loop?

File: src/sentry/utils/outcomes.py
Comment (jan-auer) [inline src/sentry/utils/outcomes.py]: Our default schema for such cache keys uses colons, for instance: `otcm:{}:{}`. Can you please change this so that it is consistent?

File: src/sentry/utils/outcomes.py
Comment (jan-auer) [inline src/sentry/utils/outcomes.py]: Please move this to where the signal is emitted instead of the outcome. Since this is only required for signals, it will be more clear if it is located there. Also, once we stop emitting signals in Sentry (by removing the code), this will automatically be gone with it.

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [inline src/sentry/outcome_consumer.py]: Can we put something with "signals" in the name, too? This consumer specifically consumes outcomes to emit signals for specific ones. Candidates I could think of:

 - `OutcomesSignalConsumer`
 - `OutcomesSignalForwarder`

Comment (RaduW) [reply src/sentry/outcome_consumer.py]: Ok, It is an outcomes consumer and signal generator, I'll think of something.

File: src/sentry/outcome_consumer.py
Comment (RaduW) [reply src/sentry/outcome_consumer.py]: ok, done

File: src/sentry/outcome_consumer.py
Comment (RaduW) [reply src/sentry/outcome_consumer.py]: yes, when there are only 2 options I prefer checking them individually.

File: src/sentry/outcome_consumer.py
Comment (RaduW) [reply src/sentry/outcome_consumer.py]: good point, done

File: src/sentry/outcome_consumer.py
Comment (RaduW) [reply src/sentry/outcome_consumer.py]: done

**Comment by jan-auer** [comment]
@JTCunning heads up: We'd like to start migrating signal emission from webs to a consumer of the `outcomes` topic. Are we stable enough, or is something blocking this from your POV?

**Comment by jan-auer** [comment]
@BYK @mattrobenolt heads up. Once this change lands, note that you might have to run this consumer in on-prem and single tenant, too.

File: src/sentry/outcome_consumer.py
Comment (jan-auer) [reply src/sentry/outcome_consumer.py]: Discussed with @tonyo. Let's stick with the more generic `OutcomesConsumer`, in case we want to add more functionality to this later.

File: src/sentry/utils/kafka.py
Comment (jan-auer) [reply src/sentry/utils/kafka.py]: This can stay, as the consumer is restarted externally.

**Comment by JTCunning** [comment]
@jan-auer: I wouldn't say that there are any blockers for this PR other than removing this option: https://github.com/getsentry/sentry/blob/master/src/sentry/utils/outcomes.py#L92.

Blockers in general though are cutting over from Web to Relay if we plan to have /stats be populated from Snuba.

If we want to unblock on that, I suggest that this consumer also be responsible for incrementing RedisTSDB keys, instead of the Kafka producer. (https://github.com/getsentry/sentry/blob/master/src/sentry/utils/outcomes.py#L56-L89)
This way, cutting over /stats to Snuba can be done separately from the Relay rollout, and can have the time required to improve the Sentry Snuba Client.

**Comment by jan-auer** [comment]
@JTCunning thanks. We're not cutting over to Relay just yet, only moving the emission of signals. We'll wait until Snuba TSDB is ready.

**Comment by jan-auer** [comment]
@tonyo note the last commit. As per James' suggestion, we're now sending outcomes unconditionally to Kafka.

File: src/sentry/utils/kafka.py
Comment (tonyo) [inline src/sentry/utils/kafka.py]: Curious: any specific reason we don't use an object attribute here?

**Review by tonyo — APPROVED** [review]
Separate 👍 for refactoring

File: src/sentry/utils/kafka.py
Comment (jan-auer) [reply src/sentry/utils/kafka.py]: The signal handler is kind of global, but this could definitely be an attribute.

**Review by tonyo — APPROVED** [review]

File: src/sentry/utils/outcomes.py
Comment (RaduW) [inline src/sentry/utils/outcomes.py]: We need to mark somewhere the outcomes coming from Sentry so that we don't double send the signal ( if we remove mark_outcome_signal from track_outcome we need to add it somewhere else in the sentry processing pipe)

**Pull request merged**
